archive/2025-11-08-code-execution-with-mcp-building-more-efficient-ai-agents-anthropic.md

```diff
@@ -0,0 +1,313 @@
+---
+date: '2025-11-08'
+description: Anthropic's latest blog post details the implementation of the Model
+  Context Protocol (MCP) to enhance AI agent efficiency through code execution. By
+  allowing agents to treat MCP servers as code APIs, developers can avoid excessive
+  token consumption tied to direct tool calls and overload on context windows. This
+  approach utilizes filesystem-like navigation to dynamically load only necessary
+  tools and process results in code, resulting in significant reductions in token
+  count—up to 98.7% in some cases. Additionally, this method facilitates better state
+  persistence, data privacy, and effective use of advanced control flows, though it
+  requires secure execution environments to manage complexity and safeguard operational
+  integrity.
+link: https://www.anthropic.com/engineering/code-execution-with-mcp
+tags:
+- AI Agents
+- Token Management
+- Efficient APIs
+- Code Execution
+- MCP
+title: 'Code execution with MCP: building more efficient AI agents \ Anthropic'
+---
+
+[Engineering at Anthropic](https://www.anthropic.com/engineering)
+
+![](https://www-cdn.anthropic.com/images/4zrzovbb/website/42f40f6fae9ec2d7cf2e5a98908a16d0216b91be-1000x1000.svg)
+
+# Code execution with MCP: Building more efficient agents
+
+Published Nov 04, 2025
+
+Direct tool calls consume context for each definition and result. Agents scale better by writing code to call tools instead. Here's how it works with MCP.
+
+[The Model Context Protocol (MCP)](https://modelcontextprotocol.io/) is an open standard for connecting AI agents to external systems. Connecting agents to tools and data traditionally requires a custom integration for each pairing, creating fragmentation and duplicated effort that makes it difficult to scale truly connected systems. MCP provides a universal protocol—developers implement MCP once in their agent and it unlocks an entire ecosystem of integrations.
+
+Since launching MCP in November 2024, adoption has been rapid: the community has built thousands of [MCP servers](https://github.com/modelcontextprotocol/servers), [SDKs](https://modelcontextprotocol.io/docs/sdk) are available for all major programming languages, and the industry has adopted MCP as the de-facto standard for connecting agents to tools and data.
+
+Today developers routinely build agents with access to hundreds or thousands of tools across dozens of MCP servers. However, as the number of connected tools grows, loading all tool definitions upfront and passing intermediate results through the context window slows down agents and increases costs.
```

Comment:
These numbers are crazy. They represent cutting edge agents, not the 99% of simple agents being built in the enterprise. This feels like a very biased view towards these apex agents.